// ng2-simple-page-scroll/src/ng2-simple-page-scroll.service.js
import { Injectable, Inject } from '@angular/core';
import { DOCUMENT } from '@angular/platform-browser';
import { SimplePageScrollConfig } from './ng2-simple-page-scroll-config';
export var SimplePageScrollService = (function () {
    function SimplePageScrollService(document) {
        this.document = document;
        this.body = document.body;
    }
    SimplePageScrollService.prototype.scrollToElement = function (targetElementOrId, pageScrollOffset) {
        var anchorTarget = this.determineElement(targetElementOrId);
        if (anchorTarget !== null) {
            setScrollTop(this.body);
            setScrollTop(this.document.documentElement);
            setScrollTop(this.document.body.parentNode);
        }
        function setScrollTop(container) {
            if (container && typeof container.scrollTop !== 'undefined') {
                container.scrollTop =
                    anchorTarget.offsetTop -
                        anchorTarget.scrollTop +
                        anchorTarget.clientTop +
                        (pageScrollOffset || SimplePageScrollConfig.defaultScrollOffset);
            }
        }
    };
    SimplePageScrollService.prototype.determineElement = function (targetElementOrId) {
        if (typeof targetElementOrId === 'string') {
            return this.document.getElementById(targetElementOrId.substr(1));
        }
        else {
            return targetElementOrId;
        }
    };
    SimplePageScrollService.decorators = [
        { type: Injectable },
    ];
    /** @nocollapse */
    SimplePageScrollService.ctorParameters = [
        { type: Document, decorators: [{ type: Inject, args: [DOCUMENT,] },] },
    ];
    return SimplePageScrollService;
}());
